lib/chef/provider/aws_ebs_volume.rb

```diff
@@ -47,13 +61,30 @@ def update_aws_object(volume)
   end
 
   def destroy_aws_object(volume)
-    converge_by "delete EBS volume #{new_resource.name} in #{region}" do
-      volume.delete
+    status = volume.status
+    case status
+    when :in_use
+      attachment = volume.attachments.first
+      detach(volume, :instance => attachment.instance, :device => attachment.device)
+      delete(volume)
+    else
+      delete(volume)
     end
   end
 
   private
 
+  def desired_instance
```

Comment:
Maybe `desired_attachment`?  The `desired` simply signifies current vs desired state.  The ivar stores what machine we expect to be attached to the volume once convergence is finished.  Does that make sense?